packages/expo-modules-core/android/src/main/java/expo/modules/adapters/react/apploader/RNHeadlessAppLoader.kt

```diff
@@ -42,27 +65,60 @@ class RNHeadlessAppLoader @DoNotStrip constructor(private val context: Context)
-      android.os.Handler(context.mainLooper).post {
-        // Only destroy the `ReactInstanceManager` if it does not bind with an Activity.
-        // And The Activity would take over the ownership of `ReactInstanceManager`.
-        // This case happens when a user clicks a background task triggered notification immediately.
-        if (appRecord.lifecycleState == LifecycleState.BEFORE_CREATE) {
-          appRecord.destroy()
+      val reactContext = appRecords[appScopeKey]!!
+      if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
+        // New architecture
+        val reactHost = (reactContext.baseContext as ReactApplication).reactHost
+        if (reactHost != null) {
+          android.os.Handler(reactContext.mainLooper).post {
+            reactHost.destroy("Closing headless task app", null)
+            HeadlessAppLoaderNotifier.notifyAppDestroyed(appScopeKey)
+            appRecords.remove(appScopeKey)
+          }
+        } else {
+          throw IllegalStateException("Your application does not have a valid reactHost")
+        }
+      } else {
+        // Old architecture
+        val reactNativeHost = (reactContext as ReactApplication).reactNativeHost
+        if (reactNativeHost.hasInstance()) {
+          val reactInstanceManager: ReactInstanceManager = reactNativeHost.reactInstanceManager
+          android.os.Handler(reactContext.mainLooper).post {
+            // Only destroy the `ReactInstanceManager` if it does not bind with an Activity.
+            // And The Activity would take over the ownership of `ReactInstanceManager`.
+            // This case happens when a user clicks a background task triggered notification immediately.
+            if (reactInstanceManager.lifecycleState == LifecycleState.BEFORE_CREATE) {
+              reactInstanceManager.destroy()
+            }
+            HeadlessAppLoaderNotifier.notifyAppDestroyed(appScopeKey)
+            appRecords.remove(appScopeKey)
+          }
         }
-        HeadlessAppLoaderNotifier.notifyAppDestroyed(appScopeKey)
-        appRecords.remove(appScopeKey)
       }
       true
     } else {
       false
     }
   }
 
-  override fun isRunning(appScopeKey: String?): Boolean =
-    appRecords.contains(appScopeKey) && appRecords[appScopeKey]!!.hasStartedCreatingInitialContext()
+  override fun isRunning(appScopeKey: String?): Boolean {
+    if (appRecords.contains(appScopeKey)) {
+      val reactContext = appRecords[appScopeKey]!!
+      if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
+        // New architecture - We can return true since the fact that we have a reactContext
+        // means that we've already called start on the reactHost
+        return true
+      } else {
+        // Old architecture
+        val reactNativeHost = (reactContext.baseContext as ReactApplication).reactNativeHost
+        return reactNativeHost.reactInstanceManager.hasStartedCreatingInitialContext()
+      }
+    }
+    return false
+  }
```

Comment:
```suggestion
    val reactContext = appRecords[appScopeKey] ?: return false
    if (BuildConfig.IS_NEW_ARCHITECTURE_ENABLED) {
      // New architecture - We can return true since the fact that we have a reactContext
      // means that we've already called start on the reactHost
      return true
    } else {
      // Old architecture
      val reactNativeHost = (reactContext.baseContext as ReactApplication).reactNativeHost
      return reactNativeHost.reactInstanceManager.hasStartedCreatingInitialContext()
    }
  }
```

also refactor to remove dangerous `!!`